target_chains/cosmwasm/tools/src/pipeline.ts

```diff
@@ -0,0 +1,181 @@
+        // past results will all be fulfilled and the pipeline will make sure of that
+        if (
+          result === undefined ||
+          (result !== undefined && result.status === "rejected")
+        ) {
+          throw new Error(
+            `${this.id}: Stage id seems to be invalid: ${stageId}`
+          );
+        }
+        return result.result;
+      };
+      try {
+        // wrapping result
+        const result = await executor(getResultOfPastStage);
+        return {
+          status: "fulfilled",
+          result,
+        };
+      } catch (e) {
+        return {
+          status: "rejected",
+          reason: e,
+        };
+      }
+    };
+  }
+
+  private async processStage(stage: Stage): Promise<boolean> {
+    // Here we will check if there is a past result that has been fulfilled
+    // If yes, we are not going to process any further
+    let currentResult = this.pipelineStore.getStageState<StageResult>(stage.id);
+    if (currentResult !== undefined && currentResult.status === "fulfilled")
+      return true;
+
+    // Else we will process the new stage and store the result
+    const newResult = await this.stageExecutorWrapper(stage.executor)();
+    this.pipelineStore.setStageState(stage.id, newResult);
+
+    if (newResult.status === "fulfilled") return true;
+
+    // Some steps can fail due to some one time errors like API issues
+    // This allows the user to re run this particular stage
+    const rerun =
+      (await readLineAsync(
+        `${this.id}: Some steps of stage: ${stage.id} failed. \n Do you want to rerun? (y) `
+      )) === "y";
+
+    if (rerun) return this.processStage(stage);
+    else return false;
+  }
+
+  async run() {
+    console.log("Running pipeline with id: ", this.id);
+    for (let stage of this.stages) {
+      console.log(`${this.id}: Running stage with id: ${stage.id}`);
+
+      // This method is only going to process stage if all the past ones have been fulfilled
+      let fulfilled = await this.processStage(stage);
+      if (fulfilled === false) break;
+    }
+
+    // store the whole processing locally
+    this.pipelineStore.commit();
+  }
+}
+
+type StoreStructure = {
+  [stageId: string]: any;
+};
+// PipelineStore helps in getting and setting the state locally
+// It manipulates data in-memory and once the consumer has finished manipulating it
+// They need to commit the data to permanent storage using the commit method
+class PipelineStore {
+  private readonly store: StoreStructure;
+
+  constructor(private readonly filePath: string) {
+    if (!existsSync(this.filePath)) {
+      this.store = {};
+      return;
+    }
+
+    this.store = JSON.parse(readFileSync(this.filePath).toString());
+  }
+
+  // It gets the latest state for the given stage
+  // the state after the last operation
+  // if there is no stage stage, in case it was no process it will return undefined.
+  getStageState<T = any>(stageId: string): T | undefined {
```

Comment:
maybe comment that the caller can basically pick the T and it'll cast it. It's the callers responsibility to know what the expected type is.